lock-api-objects/src/main/java/com/palantir/lock/watch/LockEvent.java

```diff
@@ -52,6 +66,63 @@ public <T> T accept(Visitor<T> visitor) {
         return visitor.visit(this);
     }
 
+    @JsonCreator
+    public static LockEvent create(
+            @JsonProperty("sequence") long sequence,
+            @JsonProperty("lockDescriptors") List<LockDescriptor> lockDescriptors,
+            @JsonProperty("lockToken") LockToken lockToken,
+            @JsonProperty("metadata") Optional<WireLockRequestMetadata> optWireMetadata) {
+        Optional<LockRequestMetadata> metadata = optWireMetadata.map(wireMetadata -> {
+            KeyListChecksum checksum = KeyListChecksum.of(
+                    ChecksumType.valueOf(wireMetadata.lockListChecksum().typeId()),
+                    wireMetadata.lockListChecksum().value());
+            IndexEncodingResult<LockDescriptor, ChangeMetadata> indexEncodingResult =
+                    IndexEncodingResult.<LockDescriptor, ChangeMetadata>builder()
+                            .keyList(lockDescriptors)
+                            .indexToValue(wireMetadata.indexToChangeMetadata())
+                            .keyListChecksum(checksum)
+                            .build();
+            Map<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadata =
+                    IndexEncodingUtils.decode(indexEncodingResult, Function.identity());
+            return LockRequestMetadata.of(lockDescriptorToChangeMetadata);
+        });
+        return ImmutableLockEvent.builder()
+                .sequence(sequence)
+                .lockDescriptors(ImmutableSet.copyOf(lockDescriptors))
+                .lockToken(lockToken)
+                .metadata(metadata)
+                .build();
+    }
+
+    @Unsafe
+    @JsonValue
+    public WireLockEvent toWireLockEvent() {
+        Optional<IndexEncodingResult<LockDescriptor, ChangeMetadata>> optIndexEncodingResult = metadata()
+                .map(metadata -> IndexEncodingUtils.encode(
+                        lockDescriptors(),
+                        metadata.lockDescriptorToChangeMetadata(),
+                        Function.identity(),
+                        DEFAULT_CHECKSUM_TYPE));
+        // If the lock event has metadata, we need to use the key list that was used to encode the metadata.
+        // Otherwise, any ordering of the lock descriptors is fine.
+        List<LockDescriptor> orderedLocks = optIndexEncodingResult
+                .map(IndexEncodingResult::keyList)
+                .orElseGet(() -> ImmutableList.copyOf(lockDescriptors()));
```

Comment:
Fortunately this (likely, since it's an implementation detail) won't be as expensive as it appears - ImmutableLockEvent has an ImmutableSet, and ImmutableList.copyOf optimises to avoid copying from an immutable collection where it can.

Reply:
Cool